### ISSUE TITLE
[Device MPS] Error: buffer is not large enough. Must be 19200 bytes

### ISSUE THREAD

**Issue by crayon7442** [pytorch/pytorch#78916]
### 🐛 Describe the bug

Hello to everyone,
I am trying to use the torch nightly version which supports the M1 drivers.

Here is my machine's  spec : Apple M1 Pro 14'' version (16GB Ram, 1TB SSD, 8 CPU cores)

I have tried to train object detecting with yolov5s model from the 241x241 dataset images, but failed with this error..

The error : 
```
> python3 train.py --device mps --img 416 --batch 16 --epochs 50 --data /oo/ooo/content/dataset/data.yaml --cfg ./models/yolov5s.yaml --weights yolov5s.pt --name hat_yolov5s_results

train: weights=yolov5s.pt, cfg=./models/yolov5s.yaml, data=/oo/ooo/content/dataset/data.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=50, batch_size=16, imgsz=416, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, bucket=, cache=None, image_weights=False, device=mps, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=runs/train, name=hat_yolov5s_results, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
github: up to date with https://github.com/ultralytics/yolov5 ✅
YOLOv5 🚀 v6.1-244-g47233e1 Python-3.10.4 torch-1.13.0.dev20220605 MPS

hyperparameters: lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
Weights & Biases: run 'pip install wandb' to automatically track and visualize YOLOv5 🚀 runs (RECOMMENDED)
TensorBoard: Start with 'tensorboard --logdir runs/train', view at http://localhost:6006/
Overriding model.yaml nc=80 with nc=3

                 from  n    params  module                                  arguments                     
  0                -1  1      3520  models.common.Conv                      [3, 32, 6, 2, 2]              
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     18816  models.common.C3                        [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  2    115712  models.common.C3                        [128, 128, 2]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  3    625152  models.common.C3                        [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1   1182720  models.common.C3                        [512, 512, 1]                 
  9                -1  1    656896  models.common.SPPF                      [512, 512, 5]                 
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]          
 24      [17, 20, 23]  1     21576  models.yolo.Detect                      [3, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
YOLOv5s summary: 270 layers, 7027720 parameters, 7027720 gradients

Transferred 342/349 items from yolov5s.pt
/AppleInternal/Library/BuildRoots/bxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:782: failed assertion `[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 19200 bytes
'
[1]    12846 abort      python3 train.py --device mps --img 416 --batch 16 --epochs 50 --data  --cfg 
```
Can you resolve this error?

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0.dev20220605
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.0rc2
[pip3] torch==1.13.0.dev20220605
[pip3] torchvision==0.14.0.dev20220605
[conda] numpy                     1.23.0rc2                pypi_0    pypi
[conda] numpy-base                1.22.3          py310h5e3e9f0_0  
[conda] torch                     1.13.0.dev20220605          pypi_0    pypi
[conda] torchvision               0.14.0.dev20220605          pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @kulinseth @albanD

**Comment by kulinseth** [comment]
Hi @crayon7442 , thanks for filing the issue.
Can you please provide steps to get the model and reproduce this crash ? Are there any python packages or other things we need to download to reproduce this issue ?

Kulin

**Comment by richardburleigh** [comment]
I experienced the same crash when attempting to train GANs, such as [Stylegan XL](https://github.com/autonomousvision/stylegan_xl):

```
Running MPSHeapAllocatorImpl::Malloc
Size: 10616832 Alloc Size: 10616832

Running MPSHeapAllocatorImpl::get_free_buffer

Running MPSHeapAllocatorImpl::alloc_buffer

Running MPSHeapAllocatorImpl::get_free_heap Allocated private buffer #895 with aligned size 10.12 MB (requested size: 10.12 MB, heap size: 6.69 MB, total allocated: 973.95 MB)

Running MPSHeapAllocatorImpl::Malloc
Size: 10616832 Alloc Size: 10616832

Running MPSHeapAllocatorImpl::get_free_buffer

Running MPSHeapAllocatorImpl::alloc_buffer

Running MPSHeapAllocatorImpl::get_free_heap
Allocated large private heap of size 12.00 MB (#heaps: 35, free memory: 20.33 GB)
Allocated private buffer #896 with aligned size 10.12 MB (requested size: 10.12 MB, heap size: 1.88 MB, total allocated: 984.07 MB)
/AppleInternal/Library/BuildRoots/8d3bda53-8d9c-11ec-abd7-fa6a1964e34e/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:782: failed assertion `[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 10616832 bytes
```

And in [GP-UNIT](https://github.com/williamyang1991/GP-UNIT):

```
Running MPSHeapAllocatorImpl::get_free_buffer
Reusing private buffer #678 with aligned size 2.00 MB (requested size: 1.50 MB)

Running MPSHeapAllocatorImpl::Malloc
Size: 1572864 Alloc Size: 1572864

Running MPSHeapAllocatorImpl::get_free_buffer
Reusing private buffer #526 with aligned size 2.00 MB (requested size: 1.50 MB)

Running MPSHeapAllocatorImpl::Malloc
Size: 1572864 Alloc Size: 1572864

Running MPSHeapAllocatorImpl::get_free_buffer
Reusing private buffer #525 with aligned size 2.00 MB (requested size: 1.50 MB)

Running MPSHeapAllocatorImpl::Malloc
Size: 1572864 Alloc Size: 1572864

Running MPSHeapAllocatorImpl::get_free_buffer
Reusing private buffer #529 with aligned size 2.00 MB (requested size: 1.50 MB)

Running MPSHeapAllocatorImpl::Malloc
Size: 1572864 Alloc Size: 1572864

Running MPSHeapAllocatorImpl::get_free_buffer
Reusing private buffer #530 with aligned size 2.00 MB (requested size: 1.50 MB)
/AppleInternal/Library/BuildRoots/8d3bda53-8d9c-11ec-abd7-fa6a1964e34e/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:782: failed assertion `[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 1572864 bytes
```

**Comment by giuseppebrb** [comment]
Same issue using: Python-3.9.12 torch-1.13.0.dev20220612

**Comment by giuseppebrb** [comment]
Bug it's still happening even after upgrading to the latest nightly 1.13.0.dev20220614.

YOLOv5 🚀 2022-6-15 Python-3.9.12 torch-1.13.0.dev20220614 MPS

I'm training a YOLOv5 model and this is the outcome:
```
/AppleInternal/Library/BuildRoots/b6051351-c030-11ec-96e9-3e7866fcf3a1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:782: failed assertion `[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 19200 bytes
```

here's how trigger training
```
python yolov5/train.py --img 512 --batch 32 --epochs 5 --data '../PATH_TO_MY_YAML_FILE.yaml' --weights yolov5/yolov5s.pt --cache --device mps
```

**Comment by DavidNovikov** [comment]
Any updates? I'm running into the same error

**Comment by kulinseth** [comment]
@crayon7442 , can you please check with latest nightly if this issue still exists?

**Comment by bennettwbrown** [comment]
Experiencing this issue on pytorch 1.12, any work arounds or solutions?

**Comment by DenisVieriu97** [comment]
> Experiencing this issue on pytorch 1.12, any work arounds or solutions?

@bennettwbrown this is fixed in pytorch 1.13 (nightly builds). Please give a try to that and let me know if you still any issues.

**Comment by kulinseth** [comment]
Not heard any update on the issue. Closing it now. Please re-open the issue if the original problem exists or create a new issue if its a different network or parameters to yolov5
